_posts/2018-08-22-optimize-performance-in-docker.md

```diff
@@ -0,0 +1,487 @@
+secondary config      (readonly)    /etc/ccache.conf
+stats zero time                     Wed Aug 22 23:29:32 2018
+cache hit (direct)                  3557
+cache hit (preprocessed)             229
+cache miss                          3767
+cache hit rate                     50.13 %
+called for link                      266
+called for preprocessing            2945
+compiler produced empty output       122
+compile failed                        12
+preprocessor error                    14
+bad compiler arguments                14
+unsupported source language            4
+autoconf compile/link                 64
+unsupported compiler option            4
+output to stdout                       8
+no input file                       8811
+cleanups performed                     0
+files in cache                      9023
+cache size                         204.4 MB
+max cache size                       5.0 GB
+```
+
+I'm not `ccache` expert and cannot explain all results e.g. why hit rate is so
+low, when we compile the same code?
+
+To conclude, we can gain even 30% with hot cache. Biggest gain we have when
+using multithreading, but this highly depends on CPU, in my case I had 8 jobs
+run simultaneously and gain was 40% in compilation time.
+
+# apt-cacher-ng
+
+There 2 use case for `apt-cacher-ng` in our workflows. One is Docker build
+time, which can be time-consuming since all packages and its dependencies are
+installed in the base image. Second is runtime, when you need some package that
+may have extensive dependencies e.g. `xen-systema-amd64`.
+
+First, let's setup `apt-cacher-ng`. Some guide may be found in [Docker documentation](https://docs.docker.com/engine/examples/apt-cacher-ng/), but we will modify it a little bit.
+
+Ideally, we would like to use `docker compose` to set up `apt-cacher-ng`
+container whenever it is not set, or have dedicated VM which serves this
+purpose. In this post, we consider local cache. Dockerfile may look like this:
+
+```
+FROM        ubuntu
+
+RUN     apt-get update && apt-get install -y apt-cacher-ng
+
+EXPOSE      3142
+CMD     chmod 777 /var/cache/apt-cacher-ng && /etc/init.d/apt-cacher-ng start && tail -f /var/log/apt-cacher-ng/*
+```
+
+Build and run:
+
+```
+docker build -t apt-cacher .
+docker run -d -p 3142:3142 -v $PWD/apt_cache:/var/cache/apt-cacher-ng --name cacher-container apt-cacher
+docker logs -f cacher-container
+```
+
+The output should look like this:
+
+```
+* Starting apt-cacher-ng apt-cacher-ng
+WARNING: No configuration was read from file:sfnet_mirrors
+   ...done.
+   ==> /var/log/apt-cacher-ng/apt-cacher.err <==
+
+   ==> /var/log/apt-cacher-ng/apt-cacher.log <==
+```
+
+We should also see that cacher listens on port `3142`:
+
+```
+[16:40:01] pietrushnic:~ $ netstat -an |grep 3142
+tcp6       0      0 :::3142                 :::*                    LISTEN
+```
+
+Dockerfile should contain following environment variable:
```

Comment:
This context switch is too sudden, it is hard to distinguish if it still relates to apt-cacher Dockerfile or target xen-docker Dockerfile. Maybe start sentence with: `Now our target xen-docker Dockerfile should contain ...`.